### PR TITLE
Parenting tools

### PR DESCRIPTION
**Pull request by davidlatwe**
### Problem

Avalon tools like Loader, Creator, Scene Manager, Context Manager, Workfile were not parented in Houdini and Nuke's main window.

And the Workfile app's `show()` wasn't accept `parent` arg at all.

### Changes

* Implemented main window getter function for Nuke **(Update: Also Houdini and Maya)**
* Add `parent` flag to `avalon.tools.workfiles.show()`
* Parenting tools for Nuke and Houdini
* Fix tools style after parented in Houdini

---

While I was working on Houdini, I encountered two problems:
1. The Qt style sheet was overridden by Houdini.
2. Calling Context Manager raised `Segmentation fault` fatal error.

On problem 1, the Loader has become:

<details>
<summary> See Image </summary>

image

</details>

And the Context Manager:

<details>
<summary> See Image </summary>

image

</details>

Which looks cool but I love Avalon flavor more.
Interestingly, Creator and Scene Manager did not have this issue, not sure why :(
> Edit: :point_up:  Actually they do, see comments below.

<details>
<summary> See Image </summary>

image

</details>

On problem 2, it was because of directly connecting `QtWidgets.QTreeView.selectionModel()` to the handler, and leads to a `Segmentation fault` error and crash Houdini.

The only useful info I could find was this [tweet](https://twitter.com/yantor3d/status/1161869142971961344?s=20).

### CODE REVIEW

File: avalon/nuke/pipeline.py [hunk 1]
```diff
@@ -215,6 +220,16 @@ def find_host_config(config):
...
 
 
+def get_main_window():
+    """Acquire Nuke's main window"""
+    if self._parent is None:
+        top_widgets = QtWidgets.QApplication.topLevelWidgets()
+        main_window = next(widget for widget in top_widgets if
+                           widget.isWindow() and not widget.isHidden())
+        self._parent = main_window
+    return self._parent
```
Comment (davidlatwe) [inline avalon/nuke/pipeline.py]: This was bold, but there was no obvious `QWidget` object name in Nuke. :(

File: avalon/tools/loader/app.py [hunk 2]
```diff
@@ -421,8 +421,8 @@ def show(debug=False, parent=None, use_context=False):
         lib.refresh_group_config_cache()
 
         window = Window(parent)
-        window.setStyleSheet(style.load_stylesheet())
         window.show()
+        window.setStyleSheet(style.load_stylesheet())
```
Comment (davidlatwe) [inline avalon/tools/loader/app.py]: This resolve the problem that loosing style sheet in Houdini.

File: avalon/tools/contextmanager/app.py [hunk 3]
```diff
@@ -42,6 +42,7 @@ def __init__(self, parent=None):
         task_view.setIndentation(0)
         task_model = TasksModel()
         task_view.setModel(task_model)
+        task_view_selection = task_view.selectionModel()
```
Comment (davidlatwe) [inline avalon/tools/contextmanager/app.py]: By keeping the `QItemSelectionModel` object before connecting signal, avoids the `Segmentation fault` crash in Houdini which described above.

**Comment by davidlatwe** [comment]
Thanks @tokejepsen ❤️ 
For testing, hope that @BigRoy can have a test on Houdini, and @jasperges or @jezscha  able to test on Nuke ☺️ 

Also, I don’t have fusion here, so the tools still unparented in fusion.

File: avalon/nuke/pipeline.py (hunk 1 above)
Comment (BigRoy) [reply avalon/nuke/pipeline.py]: I did find [this](https://www.mail-archive.com/nuke-python@support.thefoundry.co.uk/msg04726.html) which could be refactored to this:
```python
top_widgets = QtWidgets.QApplication.topLevelWidgets()
name  = "Foundry::UI::DockMainWindow"
main_window = next(widget for widget in top_widgets 
                   if widget.inherits('QMainWindow') and
                   if widget.metaObject().className() == name)
```

Not sure if that works though.

Comment (davidlatwe) [reply avalon/nuke/pipeline.py]: Awesome ! This works !!

**Comment by BigRoy** [comment]
> Also, I don’t have fusion here, so the tools still unparented in fusion.

There's no Qt in Fusion so I think that will have to remain as is unfortunately.

> For testing, hope that @BigRoy can have a test on Houdini

Very nice! I wonder how I missed doing this at the time. Anyway, nice work! 

### Houdini avalon tools font difference

I did notice the styling felt a bit off as it seems to be using the Houdini font as opposed or something? For example:

afbeelding

I wonder if there's a way to patch that up too.

### Houdini Work files error

The Work files tool however pops up this error on opening:
```
Traceback (most recent call last):
  File "D:\git_pipeline\avalon\development\git\core\avalon\tools\workfiles\app.py", line 344, in <lambda>
    QtCore.QTimer.singleShot(100, lambda: self.list.scrollToItem(item))
RuntimeError: Internal C++ object (PySide2.QtWidgets.QListWidgetItem) already deleted.
```

Additionally it seems that the Work Files tool does not have the Avalon styling whatsoever. Or does it look so different solely due to the font being so different?

afbeelding

**Comment by davidlatwe** [comment]
> Maybe we should try and just apply the stylesheet globally in Houdini? Sounds a bit hacky though.

Haha, and it says [here](https://github.com/shotgunsoftware/tk-houdini/blob/5df668d04748cff5d9ccfb56c3d3e237593e8751/engine.py#L247)..
 `NOTE: Except for 16+. It's no longer safe and causes lots of styling problems in Houdini's UI globally.`

So hacky indeed !

Anyway, I finally resolved the font size and `QSplitter` issue by adding some entry in `style.qss`, the Loader and Creator tools should now look almost the same as in other DCC.

Here's the before and after:

#### before

image

Notice that the splitter in Loader is a thick gray line even mouse is not hover on top of them.

#### after

image

:tada:

Furthermore, I also moved the line `window.setStyleSheet` after `window.show` in every tool, so their appearance should all aligned now.

File: avalon/tools/workfiles/app.py [hunk 4]
```diff
@@ -337,11 +337,14 @@ def refresh(self):
...
-            item = self.list.item(modified.index(max(modified)))
+            index = modified.index(max(modified))
+            item = self.list.item(index)
             item.setSelected(True)
 
             # Scroll list so item is visible
-            QtCore.QTimer.singleShot(100, lambda: self.list.scrollToItem(item))
+            def scroll_to_item():
+                self.list.scrollToItem(self.list.item(index))
+            QtCore.QTimer.singleShot(100, scroll_to_item)
```
[the remark this answers is not in the document]
Comment (davidlatwe) [reply avalon/tools/workfiles/app.py]: Interestingly, if I keep `hou.ui.mainQtWindow()` inside `hou.session`, the issue "QListWidgetItem already deleted" happened in Workfile app will not happen. But the `Segmentation fault` error caused by `QTreeView.selectionModel()` signal connecting still exists.

[the remark this answers is not in the document]
Comment (davidlatwe) [reply avalon/tools/workfiles/app.py]: Implemented a much proper fix in https://github.com/getavalon/core/pull/472/commits/5907a80bcd2e0ff6457be3e0617d054cd127ebe9

**Comment by davidlatwe** [comment]
> ### Houdini Work files error
> 
> The Work files tool however pops up this error on opening:
> ```
> Traceback (most recent call last):
>   File "D:\git_pipeline\avalon\development\git\core\avalon\tools\workfiles\app.py", line 344, in <lambda>
>     QtCore.QTimer.singleShot(100, lambda: self.list.scrollToItem(item))
> RuntimeError: Internal C++ object (PySide2.QtWidgets.QListWidgetItem) already deleted.
> ```

Originally, this issue was fixed in https://github.com/getavalon/core/pull/472/commits/1b2e9add741848c3e933fe03ccd74374ca8045e3, but after a few more search, I find that keeping the Houdini's main window instance in a place which can have a longer lifespan is a better solution, which fixes in a higher scope (based on this [post](https://www.sidefx.com/forum/topic/54828/?page=1#post-246287)). So I revert the previous commit and add https://github.com/getavalon/core/pull/472/commits/5907a80bcd2e0ff6457be3e0617d054cd127ebe9.

And for removing duplicated main window getting logic plus having same function for each host, I also add `get_main_window` in `maya.pipeline` in https://github.com/getavalon/core/pull/472/commits/49687febcb7cc5072b4203f4ba6f7ecd85d95954.

**Comment by tokejepsen** [comment]
Just noticed as well that the Loader is not parent to the Maya window.

**Comment by davidlatwe** [comment]
@tokejepsen, really !? Loader did get parented in Maya here in my end. :astonished:

**Comment by tokejepsen** [comment]
> @tokejepsen, really !? Loader did get parented in Maya here in my end. 😲

Ok, I'm not on the latest core, so it could just be something with my fork. Nvm :)

**Comment by davidlatwe** [comment]
Will merge this tomorrow if no other objections :)

**Comment by mottosso** [comment]
I've scanned through the code and couldn't find anything I didn't like. Nice work. :)

**Pull request merged**